File: 02_02_return_kth_to_last.java
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        ListNode dummyHead = new ListNode(-1, head);
        ListNode leader = dummyHead;
        
        int count = 0;
        
        while(leader != null) {
            leader = leader.next;
            count++;
            if (count == n) {
                break;
            }
        }
        
        ListNode follower = dummyHead;
        
        while (leader.next != null) {
            leader = leader.next;
            follower = follower.next;
        }
        
        follower.next = follower.next.next;
        
        return dummyHead.next;
    }
}

File: 02_02_Return_Kth_To_Last.cpp
/** Solution 1**/
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        auto * dummyHead = new ListNode(0);
        
        auto count = 0;

        for (auto * current = head; current != nullptr; current = current->next)
            count++;
        
        count -= n;

        dummyHead->next = head;
        
        auto * current = dummyHead;
        
        while(count--)
            current = current->next;
        
        current->next = current->next->next;
        
        return dummyHead->next;
    }
};

/** Solution 2 **/
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {

        auto * dummyHead = new ListNode;
        dummyHead->next = head;
        
        auto * fast = dummyHead;
        auto * slow = dummyHead;
        
        for(auto i = 0; i < n; i++)
            fast = fast->next;
        
        while(fast->next != nullptr)
        {
            slow = slow->next;
            fast = fast->next;
        }
        
        slow->next = slow->next->next;
        
        return dummyHead->next;
    }
};
